### PR TITLE
Deny global_asm in deny(unsafe_code)

### LINKED ISSUE #103078: deny(unsafe_code) should prevent `global_asm!` usage.
Similar to how deny(unsafe_code) prevents `#[no_mangle]` from being used, it should also prevent `global_asm!` blocks from being allowed. If misused, they can do any number of strange and horrible thing simply by being within the crate.

### PR DESCRIPTION
**Pull request by ghost**
Fixes #103078

### CODE REVIEW

**Comment by Nilstrieb** [comment]
You have your test binary checked in

**Review by fee1-dead — CHANGES_REQUESTED** [review]
Please add a ui test, also please squash the commit so that the binary does not appear in the history.

File: compiler/rustc_middle/src/lint.rs [hunk 1]
```diff
@@ -483,7 +483,10 @@ pub fn in_external_macro(sess: &Session, span: Span) -> bool {
             DesugaringKind::ForLoop | DesugaringKind::WhileLoop | DesugaringKind::OpaqueTy,
         ) => false,
         ExpnKind::AstPass(_) | ExpnKind::Desugaring(_) => true, // well, it's "external"
-        ExpnKind::Macro(MacroKind::Bang, _) => {
+        ExpnKind::Macro(MacroKind::Bang, id) => {
+            if id.as_str() == "core::arch::global_asm" {
+                return false;
+            }
```
Comment (fee1-dead) [inline compiler/rustc_middle/src/lint.rs:489]: Please add a comment explaining that this is needed such that the `unsafe_code` lint works correctly.

Also, what if an external macro generates `global_asm`? Would that be denied by the compiler?

**Comment by gimbles** [comment]
OH NO

**Comment by gimbles** [comment]
I messed up a rebase I seriously didn't mean to ping everyone 😓

**Pull request closed without merging**